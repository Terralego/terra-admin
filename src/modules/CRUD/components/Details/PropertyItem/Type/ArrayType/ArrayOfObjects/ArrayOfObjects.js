import React from 'react';
import PropTypes from 'prop-types';
import { getObjectOrderedValue, isTableObject } from '../../../../../../services/utils';

import Type from '../../Type';

const ArrayOfObjects = ({
  display_value: displayValue,
  schema: { items: schemaItems },
  ui_schema: { items: uiSchemaItems = {} } = {},
}) => {
  const orderedValue = displayValue.map(value => getObjectOrderedValue(value, uiSchemaItems['ui:order']));

  if (!isTableObject(orderedValue)) {
    return displayValue.map((value, i) => (
      <Type
        // eslint-disable-next-line react/no-array-index-key
        key={i}
        display_value={value}
        schema={schemaItems}
        ui_schema={uiSchemaItems}
        type={schemaItems.type}
      />
    ));
  }

  const { properties = {}, type } = schemaItems;
  const columns = Object.entries(orderedValue[0]);
  return (
    <div className="details__table">
      <table>
        <thead>
          <tr>
            {columns.map(([key, value]) => (
              <th key={key}>{properties[key].title || value}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {orderedValue.map((row, i) => (
          // eslint-disable-next-line react/no-array-index-key
            <tr key={i}>
              {Object.entries(row).map(([key, value]) => (
                <td key={key}>
                  <Type
                    display_value={value}
                    schema={properties[key]}
                    ui_schema={uiSchemaItems[key]}
                    type={type[key]}
                  />
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

ArrayOfObjects.propTypes = {
  display_value: PropTypes.arrayOf(PropTypes.shape({})),
  schema: PropTypes.shape({
    items: PropTypes.shape({
      type: PropTypes.string,
    }),
  }),
  ui_schema: PropTypes.shape({
    items: PropTypes.shape({
      'ui:order': PropTypes.array,
    }),
  }),
};

ArrayOfObjects.defaultProps = {
  display_value: [],
  schema: {
    items: {
      type: undefined,
    },
  },
  ui_schema: {
    items: {
      'ui:order': [],
    },
  },
};

export default ArrayOfObjects;
