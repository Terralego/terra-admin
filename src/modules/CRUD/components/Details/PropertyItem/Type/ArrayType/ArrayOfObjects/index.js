import ArrayOfObjects from './ArrayOfObjects';

export default ArrayOfObjects;
