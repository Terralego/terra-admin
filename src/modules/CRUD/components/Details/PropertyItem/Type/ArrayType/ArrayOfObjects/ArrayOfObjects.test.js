
import React from 'react';
import renderer from 'react-test-renderer';
import ArrayOfObjects from './ArrayOfObjects';

jest.mock('../../Type', () => props => <div {...props} />);

const arrayObjectsProps = {
  display_value: [{
    date: '2019',
    source: 'Requalification du site',
  }, {
    date: '2018',
    source: 'Réfection toiture',
  }],
  schema: {
    items: {
      properties: {
        date: {
          type: 'string',
          title: 'Date',
          format: 'date',
        },
        source: {
          type: 'string',
          title: 'Source',
        },
      },
      type: 'object',
    },
  },
};

describe('should render correctly', () => {
  it('array of objects as table', () => {
    const tree = renderer.create((
      <ArrayOfObjects {...arrayObjectsProps} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('array of objects', () => {
    const tree = renderer.create((
      <ArrayOfObjects
        {...arrayObjectsProps}
        display_value={[{
          date: '2019',
          source: 'Requalification du site',
        }, {
          date: '2018',
          source: 'Réfection toiture',
          foo: 'foo',
        }]}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
