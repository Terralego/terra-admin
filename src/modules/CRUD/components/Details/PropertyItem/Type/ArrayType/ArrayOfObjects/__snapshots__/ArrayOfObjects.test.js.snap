// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly array of objects 1`] = `
Array [
  <div
    display_value={
      Object {
        "date": "2019",
        "source": "Requalification du site",
      }
    }
    schema={
      Object {
        "properties": Object {
          "date": Object {
            "format": "date",
            "title": "Date",
            "type": "string",
          },
          "source": Object {
            "title": "Source",
            "type": "string",
          },
        },
        "type": "object",
      }
    }
    type="object"
    ui_schema={
      Object {
        "ui:order": Array [],
      }
    }
  />,
  <div
    display_value={
      Object {
        "date": "2018",
        "foo": "foo",
        "source": "Réfection toiture",
      }
    }
    schema={
      Object {
        "properties": Object {
          "date": Object {
            "format": "date",
            "title": "Date",
            "type": "string",
          },
          "source": Object {
            "title": "Source",
            "type": "string",
          },
        },
        "type": "object",
      }
    }
    type="object"
    ui_schema={
      Object {
        "ui:order": Array [],
      }
    }
  />,
]
`;

exports[`should render correctly array of objects as table 1`] = `
<div
  className="details__table"
>
  <table>
    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Source
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            display_value="2019"
            schema={
              Object {
                "format": "date",
                "title": "Date",
                "type": "string",
              }
            }
          />
        </td>
        <td>
          <div
            display_value="Requalification du site"
            schema={
              Object {
                "title": "Source",
                "type": "string",
              }
            }
          />
        </td>
      </tr>
      <tr>
        <td>
          <div
            display_value="2018"
            schema={
              Object {
                "format": "date",
                "title": "Date",
                "type": "string",
              }
            }
          />
        </td>
        <td>
          <div
            display_value="Réfection toiture"
            schema={
              Object {
                "title": "Source",
                "type": "string",
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
